### DISCUSSION TITLE
Querge Surge API integration

### DISCUSSION THREAD

**Discussion by tkumar911921** [community/community#141675] [Apps, API and Webhooks]
### Select Topic Area

Question

### Body

Hi Team,
I have one question; can we integrate Query Surge API in GitHub actions for the CICD to automate the unit test process. It would be great if can provide some info. Much appreciated 😊

**Comment by MorelliniMatteo** [comment]
Yes, you can integrate QuerySurge API with GitHub Actions for your CI/CD pipeline to automate the unit test process! Here’s a general guide on how you can achieve that:

Steps to Integrate QuerySurge API in GitHub Actions:
1. Set Up QuerySurge API:
Make sure that your QuerySurge API is accessible and set up correctly.
QuerySurge provides REST API endpoints to interact with test runs, project management, and result retrieval, which can be used to trigger and monitor tests during your CI/CD pipeline.
2. Create a GitHub Actions Workflow:
You will create or modify a GitHub Actions workflow file (.yml) to call the QuerySurge API from your CI pipeline. Here’s an example of how you can structure the workflow to integrate the QuerySurge API.

3. Example GitHub Actions Workflow:
yaml
Copia codice
name: CI/CD with QuerySurge

on:
  push:
    branches:
      - main
  pull_request:

jobs:
  run-querysurge-tests:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout code
        uses: actions/checkout@v3

      - name: Run QuerySurge Tests
        id: querysurge
        run: |
          # Step 1: Trigger a QuerySurge test run via API
          curl -X POST https://your-querysurge-api-endpoint/api/test-runs \
          -H "Content-Type: application/json" \
          -H "Authorization: Bearer ${{ secrets.QUERYSURGE_API_TOKEN }}" \
          -d '{
              "testSuiteId": "<YourTestSuiteId>",
              "projectId": "<YourProjectId>"
              }'

          # Step 2: Poll the API until the test results are ready
          # You can implement a loop here to check the status of the test run
          # Example pseudo-code:
          # while [ status != "completed" ]; do
          #   sleep 30
          #   status=$(curl -s https://your-querysurge-api-endpoint/api/test-runs/status)
          # done

      - name: Fetch Test Results
        run: |
          # Fetch the test results using QuerySurge API
          curl -X GET https://your-querysurge-api-endpoint/api/test-runs/<YourTestRunId>/results \
          -H "Authorization: Bearer ${{ secrets.QUERYSURGE_API_TOKEN }}" \
          -o results.json

      - name: Upload Test Results Artifact
        uses: actions/upload-artifact@v3
        with:
          name: querysurge-results
          path: results.json
4. Key Steps in the Workflow:
Checkout Code: This step ensures that your code is pulled from the GitHub repository.
Trigger QuerySurge Tests: Using the curl command, you make an API call to start a QuerySurge test run. You pass the required test suite and project IDs.
Polling the Test Status: After triggering the test, you can set up a loop to keep polling the API until the test results are available. (You can handle this step using more advanced logic in a shell script.)
Fetching Test Results: Once the tests are complete, retrieve the results via the API and store them.
Upload Results as Artifacts: This step uploads the test results as an artifact, so they can be reviewed post-run.
5. Authentication:
Store the QuerySurge API token in your GitHub repository’s Secrets for security.
Go to your GitHub repo → Settings → Secrets and Variables → Actions → New Repository Secret.
Add your QuerySurge API token here and use it in the workflow as secrets.QUERYSURGE_API_TOKEN.
6. Handling Errors:
Implement error handling in the script (e.g., fail the GitHub Actions job if the QuerySurge test run fails).
You can use exit 1 in the script to indicate failure and stop the pipeline if needed.

Please mark this answr as answered

**Comment by tkumar911921** [comment]
Thanks for the answer 👍
